# Telecom churn case study

### Business Objective: The business objective is to predict the churn in the last (i.e. the ninth) month using the data (features) from the first three months

In [1]:
# import libraries

import numpy as np
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import scale


In [2]:
# read the file

data=pd.read_csv('telecom_churn_data.csv')
data.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0            9/30/2014  197.385  214.816  213.803   21.100          NaN   
1            9/30/2014   34.047  355.074  268.321   86.285        24.11   
2            9/30/2014  167.690  189.058  210.226  290.714        11.54   
3            9/30/2014  221.338  251.102  508.054  389.500        99.91   
4            9/30/2014  261.636  309.876  238.174  163.426        50.31   

   onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0          NaN         0.00          NaN           NaN           NaN   
1        78.68         7.68        18.34         15.74         99.84   
2        55.24        37.26        74.81        143.33        220.59   
3        54.39       310.98       241.71        123.31        109.01   
4       149.44        83.89        58.78         76.96         91.88   

   offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0          0.00           NaN            NaN            NaN           0.00   
1        304.76         53.76            0.0           0.00           0.00   
2        208.36        118.91            0.0           0.00           0.00   
3         71.68        113.54            0.0          54.86          44.38   
4        124.26         45.81            0.0           0.00           0.00   

   roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0            NaN            NaN            NaN           0.00            NaN   
1           0.00            0.0           0.00           0.00           0.00   
2          38.49            0.0           0.00           0.00          70.94   
3           0.00            0.0          28.09          39.04           0.00   
4           0.00            0.0           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0               NaN               NaN              0.00               NaN   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   
3             73.68             34.81             10.61             15.49   
4             50.31            149.44             83.89             58.78   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0               NaN               NaN              0.00               NaN   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   
3            107.43             83.21             22.46             65.46   
4             67.64             91.88            124.26             37.89   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0               NaN               NaN              0.00               NaN   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   
3              1.91              0.65              

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 226 entries, mobile_number to sep_vbc_3g
dtypes: float64(179), int64(35), object(12)
memory usage: 172.4+ MB


In [4]:
data.shape

(99999, 226)

In [5]:
#store dates in a dataset and numeric ones in one

all_columns=data.columns
date_columns=[x for x in all_columns if 'date' in x]
numeric_columns=[x for x in all_columns if x not in date_columns]

### Check for Null Values

In [6]:
data.isnull().sum()

mobile_number                   0
circle_id                       0
loc_og_t2o_mou               1018
std_og_t2o_mou               1018
loc_ic_t2o_mou               1018
last_date_of_month_6            0
last_date_of_month_7          601
last_date_of_month_8         1100
last_date_of_month_9         1659
arpu_6                          0
arpu_7                          0
arpu_8                          0
arpu_9                          0
onnet_mou_6                  3937
onnet_mou_7                  3859
onnet_mou_8                  5378
onnet_mou_9                  7745
offnet_mou_6                 3937
offnet_mou_7                 3859
offnet_mou_8                 5378
offnet_mou_9                 7745
roam_ic_mou_6                3937
roam_ic_mou_7                3859
roam_ic_mou_8                5378
roam_ic_mou_9                7745
roam_og_mou_6                3937
roam_og_mou_7                3859
roam_og_mou_8                5378
roam_og_mou_9                7745
loc_og_t2t_mou

In [7]:
# lets fill the null values with 0 for numeric columns first because this implies no recharge was made

data[numeric_columns]=data[numeric_columns].fillna(0)

In [8]:
# lets check the date columns

data[date_columns].head()

last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9 date_of_last_rech_6 date_of_last_rech_7  \
0            9/30/2014           6/21/2014           7/16/2014   
1            9/30/2014           6/29/2014           7/31/2014   
2            9/30/2014           6/17/2014           7/24/2014   
3            9/30/2014           6/28/2014           7/31/2014   
4            9/30/2014           6/26/2014           7/28/2014   

  date_of_last_rech_8 date_of_last_rech_9 date_of_last_rech_data_6  \
0            8/8/2014           9/28/2014                6/21/2014   
1           8/28/2014           9/30/2014                      NaN   
2           8/14/2014           9/29/2014                      NaN   
3           8/31/2014           9/30/2014                      NaN   
4            8/9/2014           9/28/2014                 6/4/2014   

  date_of_last_rech_data_7 date_of_last_rech_data_8 date_of_last_rech_data_9  
0                7/16/2014                 8/8/2014                      NaN  
1                7/25/2014                8/10/2014                      NaN  
2                      NaN                      NaN                9/17/2014  
3                      NaN                      NaN                      NaN  
4                      NaN                      NaN                      NaN

In [9]:
# replace nan values with 

from datetime import date,datetime

data['last_date_of_month_7']=data['last_date_of_month_7'].fillna('7/31/2014')
data['last_date_of_month_8']=data['last_date_of_month_8'].fillna('8/31/2014')
data['last_date_of_month_9']=data['last_date_of_month_9'].fillna('9/30/2014')

#datetime.datetime.strptime("2013-1-25", '%Y-%m-%d').strftime('%m/%d/%y')

data[date_columns]=data[date_columns].fillna(datetime.strptime(str(date.max), '%Y-%m-%d').strftime('%m/%d/%Y'))


In [10]:
round(data[date_columns].isnull().mean(),2)

last_date_of_month_6        0.0
last_date_of_month_7        0.0
last_date_of_month_8        0.0
last_date_of_month_9        0.0
date_of_last_rech_6         0.0
date_of_last_rech_7         0.0
date_of_last_rech_8         0.0
date_of_last_rech_9         0.0
date_of_last_rech_data_6    0.0
date_of_last_rech_data_7    0.0
date_of_last_rech_data_8    0.0
date_of_last_rech_data_9    0.0
dtype: float64

**null values have been dealt with**

In [11]:
data.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0            9/30/2014  197.385  214.816  213.803   21.100         0.00   
1            9/30/2014   34.047  355.074  268.321   86.285        24.11   
2            9/30/2014  167.690  189.058  210.226  290.714        11.54   
3            9/30/2014  221.338  251.102  508.054  389.500        99.91   
4            9/30/2014  261.636  309.876  238.174  163.426        50.31   

   onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0         0.00         0.00         0.00          0.00          0.00   
1        78.68         7.68        18.34         15.74         99.84   
2        55.24        37.26        74.81        143.33        220.59   
3        54.39       310.98       241.71        123.31        109.01   
4       149.44        83.89        58.78         76.96         91.88   

   offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0          0.00          0.00            0.0           0.00           0.00   
1        304.76         53.76            0.0           0.00           0.00   
2        208.36        118.91            0.0           0.00           0.00   
3         71.68        113.54            0.0          54.86          44.38   
4        124.26         45.81            0.0           0.00           0.00   

   roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0           0.00            0.0           0.00           0.00           0.00   
1           0.00            0.0           0.00           0.00           0.00   
2          38.49            0.0           0.00           0.00          70.94   
3           0.00            0.0          28.09          39.04           0.00   
4           0.00            0.0           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0              0.00              0.00              0.00              0.00   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   
3             73.68             34.81             10.61             15.49   
4             50.31            149.44             83.89             58.78   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0              0.00              0.00              0.00              0.00   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   
3            107.43             83.21             22.46             65.46   
4             67.64             91.88            124.26             37.89   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0              0.00              0.00              0.00              0.00   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   
3              1.91              0.65              

## Derive new features

In [12]:
# Code for deriving new features

#add all total services pertaining to the specific month
data['total_service_6'] = data['night_pck_user_6']+data['monthly_2g_6']+data['sachet_2g_6']+data['monthly_3g_6']+data['sachet_3g_6']+data['fb_user_6']
data['total_service_7'] = data['night_pck_user_7']+data['monthly_2g_7']+data['sachet_2g_7']+data['monthly_3g_7']+data['sachet_3g_7']+data['fb_user_7']
data['total_service_8'] = data['night_pck_user_8']+data['monthly_2g_8']+data['sachet_2g_8']+data['monthly_3g_8']+data['sachet_3g_8']+data['fb_user_8']
data['total_service_9'] = data['night_pck_user_9']+data['monthly_2g_9']+data['sachet_2g_9']+data['monthly_3g_9']+data['sachet_3g_9']+data['fb_user_9']

#difference in outgoing calls and incoming calls on a monthly basis
data['difference_ic_og_6'] = data['total_ic_mou_6']-data['total_og_mou_6']
data['difference_ic_og_7'] = data['total_ic_mou_7']-data['total_og_mou_7']
data['difference_ic_og_8'] = data['total_ic_mou_8']-data['total_og_mou_8']
data['difference_ic_og_9'] = data['total_ic_mou_9']-data['total_og_mou_9']

#mobile internet usage in MB per month 3g/2g
data['difference_vol_3g_2g_6'] = data['vol_3g_mb_6']-data['vol_2g_mb_6']
data['difference_vol_3g_2g_7'] = data['vol_3g_mb_7']-data['vol_2g_mb_7']
data['difference_vol_3g_2g_8'] = data['vol_3g_mb_8']-data['vol_2g_mb_8']
data['difference_vol_3g_2g_9'] = data['vol_3g_mb_9']-data['vol_2g_mb_9']

#total service used while being active
data['difference_total_service_packs_good_active_phase']=data['total_service_6']+data['total_service_7']-data['total_service_8']
data['difference_ic_good_active_phase']=data['total_ic_mou_6']+data['total_ic_mou_7']-data['total_ic_mou_8']
data['difference_og_good_active_phase']=data['total_og_mou_6']+data['total_og_mou_7']-data['total_og_mou_8']

#total internet service used while being active
data['difference_2g_good_active_phase']=data['vol_2g_mb_6']+data['vol_2g_mb_7']-data['vol_2g_mb_8']
data['difference_3g_good_active_phase']=data['vol_3g_mb_6']+data['vol_3g_mb_7']-data['vol_3g_mb_8']

#total recharge amount while being active
data['difference_total_rech_amt_good_active']= data['total_rech_amt_6']+data['total_rech_amt_7']-data['total_rech_amt_8']

In [13]:
#total data recharge for month of 6 and 7
data['total_rech_data_amt_good_phase'] = data['total_rech_data_6']*data['av_rech_amt_data_6']+ data['total_rech_data_7']*data['av_rech_amt_data_7']

#total normal recharge for month of 6 and 7
data['total_rech_amt_good_phase'] = data['total_rech_amt_6']+data['total_rech_amt_7']

#total recharge = data + normal
data['total_amt_good_phase'] = data['total_rech_data_amt_good_phase']+data['total_rech_amt_good_phase']


In [14]:
#check percentage to find our high value customers which is 70% above

data['total_amt_good_phase'].describe(percentiles=[.25,.5,.70,.90,.95,.99])

count     99999.000000
mean        970.656135
std        1738.847187
min           0.000000
25%         256.000000
50%         550.000000
70%         956.000000
90%        2056.000000
95%        3024.000000
99%        6727.020000
max      122472.000000
Name: total_amt_good_phase, dtype: float64

## Filter High Value Customers (70th percentile onwards)

In [15]:
data_hvc =data[data['total_amt_good_phase']>=956.00]
data_hvc.shape

(30001, 247)

## Tag Churners

Attributes to use to tag churners

total_ic_mou_9,
total_og_mou_9,
vol_2g_mb_9,
vol_3g_mb_9.

In [16]:
data_hvc['churn']=data_hvc['total_ic_mou_9']+data_hvc['total_og_mou_9']+data_hvc['vol_2g_mb_9']+data_hvc['vol_3g_mb_9']

data_hvc['churn']=data_hvc['churn'].apply(lambda x: 1 if x==0 else 0)

In [17]:
data_hvc['churn'].value_counts()

0    27560
1     2441
Name: churn, dtype: int64

In [18]:
#churn percentage

data_hvc['churn'].value_counts()*100/data_hvc['churn'].shape[0]

0    91.863605
1     8.136395
Name: churn, dtype: float64

2441 indicates the number of people churned who are high values customers

## remove attributes of the churn phase

In [19]:
remove_att_churn=[x for x in data_hvc.columns if '_9' not in x and 'sep' not in x]


In [20]:
data_hvc=data_hvc[remove_att_churn]
data_hvc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30001 entries, 0 to 99997
Columns: 190 entries, mobile_number to churn
dtypes: float64(151), int64(30), object(9)
memory usage: 43.7+ MB


In [21]:
# check for duplicates

if len(data_hvc['mobile_number'].unique()) == len(data_hvc.index):
    print("No Duplicates. Each Mobile Number is unique.")
else:
    print("Duplicate observations present.")

No Duplicates. Each Mobile Number is unique.


## Checking for outliers

In [22]:
# Checking for outliers in the continuous variables

num_data = data_hvc[[x for x in data_hvc.columns if x not in date_columns]]

In [23]:
num_data.describe(percentiles=[0.02,.25,.5,.75,.90,.95,.99])

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   3.000100e+04    30001.0         30001.0         30001.0   
mean    7.001206e+09      109.0             0.0             0.0   
std     6.908784e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
2%      7.000049e+09      109.0             0.0             0.0   
25%     7.000609e+09      109.0             0.0             0.0   
50%     7.001203e+09      109.0             0.0             0.0   
75%     7.001804e+09      109.0             0.0             0.0   
90%     7.002165e+09      109.0             0.0             0.0   
95%     7.002285e+09      109.0             0.0             0.0   
99%     7.002386e+09      109.0             0.0             0.0   
max     7.002411e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         30001.0  30001.000000  30001.000000  30001.000000  30001.000000   
mean              0.0    558.490824    560.782203    508.597957    260.793024   
std               0.0    460.640461    479.776947    501.961981    459.644368   
min               0.0  -2258.709000  -2014.045000   -945.808000      0.000000   
2%                0.0     39.839000     43.957000      0.000000      0.000000   
25%               0.0    309.865000    309.826000    231.473000     17.080000   
50%               0.0    481.694000    480.943000    427.585000     84.580000   
75%               0.0    699.943000    698.315000    661.491000    290.440000   
90%               0.0    994.099000    995.859000    977.345000    754.160000   
95%               0.0   1240.964000   1261.272000   1255.019000   1135.440000   
99%               0.0   1985.115000   1999.500000   1986.622000   2151.740000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  30001.000000  30001.000000  30001.000000  30001.000000  30001.000000   
mean     267.819295    234.112539    373.693961    378.103169    335.077044   
std      479.993989    458.448598    482.523558    498.923555    482.062509   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
2%         0.000000      0.000000      0.000000      0.000000      0.000000   
25%       16.030000     10.390000     71.610000     69.910000     46.740000   
50%       82.810000     65.610000    222.540000    220.030000    182.790000   
75%      290.240000    239.960000    487.940000    494.010000    438.890000   
90%      784.480000    665.080000    895.830000    916.080000    823.680000   
95%     1185.790000   1074.590000   1256.610000   1272.290000   1167.540000   
99%     2201.960000   2159.110000   2326.360000   2410.890000   2193.130000   
max     8157.780000  10752.560000   8362.360000   9667.130000  14007.340000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   30001.000000   30001.000000   30001.000000   30001.000000   
mean       16.110355      12.642504      12.500551      26.571547   
std        76.302156      75.785903      74.125281     116.205525   
min         0.000000       0.000000       0.000000       0.000000   
2%          0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
90%        27.390000      14.290000      15.010000      50.430000   
95%        84.540000      55.640000      56.350000     145.410000   
99%       342.440000     280.460000     282.190000     530.710000   
max      2613.310000    3813.290000    4169.810000    3775.110000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   30001.000000   30001.000000      30001.000000      30001.000000   
mean 

Outliers are present 

### Treating the outliers

In [24]:
filter_data = data_hvc[[x for x in data_hvc.columns if x not in date_columns]]
low = .001
high = .999
quant_df = filter_data.quantile([low, high])
quant_df

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
0.001   7.000002e+09      109.0             0.0             0.0   
0.999   7.002408e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou    arpu_6    arpu_7    arpu_8  onnet_mou_6  onnet_mou_7  \
0.001             0.0    -3.000    -7.420    -5.000         0.00         0.00   
0.999             0.0  3884.712  4115.703  3976.261      4136.34      4355.98   

       onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
0.001         0.00          0.00          0.00          0.00           0.00   
0.999      4165.71       4175.83       4546.48       4472.46         920.59   

       roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  \
0.001           0.00           0.00           0.00           0.00   
0.999        1022.38         917.96        1458.89        1244.64   

       roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
0.001           0.00              0.00              0.00              0.00   
0.999        1365.11           3006.24           3098.94           2767.29   

       loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  \
0.001              0.00              0.00              0.00              0.00   
0.999           2401.48           2192.58           2099.48            257.18   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
0.001              0.00              0.00              0.00              0.00   
0.999            242.26            234.68             72.29             90.53   

       loc_og_t2c_mou_8  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
0.001              0.00          0.00          0.00          0.00   
0.999             75.43       3508.19       3604.84       3353.39   

       std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  \
0.001              0.00              0.00              0.00              0.00   
0.999           3708.53           3877.74           3779.73           3884.23   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  \
0.001              0.00              0.00              0.00              0.00   
0.999           4267.94           4240.41            180.08            183.98   

       std_og_t2f_mou_8  std_og_t2c_mou_6  std_og_t2c_mou_7  std_og_t2c_mou_8  \
0.001              0.00               0.0               0.0               0.0   
0.999            161.74               0.0               0.0               0.0   

       std_og_mou_6  std_og_mou_7  std_og_mou_8  isd_og_mou_6  isd_og_mou_7  \
0.001          0.00          0.00          0.00          0.00          0.00   
0.999       4929.71       5003.64       5269.71        301.69        302.21   

       isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  og_others_6  \
0.001          0.00          0.00          0.00          0.00         0.00   
0.999        358.88        208.46        253.73        273.46        23.53   

       og_others_7  og_others_8  total_og_mou_6  total_og_mou_7  \
0.001         0.00          0.0            0.00            0.00   
0.999         6.09          6.7         5423.59         5766.44   

       total_og_mou_8  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  loc_ic_t2t_mou_8  \
0.001            0.00              0.00              0.00              0.00   
0.999         5762.48           2182.88           2436.23           2169.71   

       loc_ic_t2m_mou_6  loc_ic_t2m_mou_7  loc_ic_t2m_mou_8  loc_ic_t2f_mou_6  \
0.001              0.00              0.00              0.00              0.00   
0.999           2225.44           2356.78           2172.51            512.79   

       loc_ic_t2f_mou_7  loc_ic_t2f_mou_8  loc_ic_mou_6  loc_ic_mou_7  \
0.001              0.00              0.00          0.00          0.00   
0.999            602.33            487.44       3075.49       3150.29   

       loc_ic_mou_8  std_ic_t2t_mou_6  std_ic_t2t_mou_7  std_ic_t2t_mou_8  

In [25]:
filter_data = filter_data.apply(lambda x: x[(x>=quant_df.loc[low,x.name]) & 
                                    (x <= quant_df.loc[high,x.name])], axis=0)

In [26]:
#merge the dataframes

data_hvc = pd.concat([data_hvc.loc[:, ~data_hvc.columns.isin([x for x in data_hvc.columns if x not in date_columns])], filter_data], axis=1)


In [27]:
#Dropping NA values

data_hvc.dropna(inplace=True)
data_hvc.shape

(27861, 190)

# EDA

In [28]:
(data_hvc.corr())

mobile_number  circle_id  \
mobile_number                                          1.000000        NaN   
circle_id                                                   NaN        NaN   
loc_og_t2o_mou                                              NaN        NaN   
std_og_t2o_mou                                              NaN        NaN   
loc_ic_t2o_mou                                              NaN        NaN   
arpu_6                                                 0.040582        NaN   
arpu_7                                                 0.042131        NaN   
arpu_8                                                 0.051268        NaN   
onnet_mou_6                                            0.008680        NaN   
onnet_mou_7                                            0.005527        NaN   
onnet_mou_8                                            0.008375        NaN   
offnet_mou_6                                           0.023314        NaN   
offnet_mou_7                                           0.013612        NaN   
offnet_mou_8                                           0.022316        NaN   
roam_ic_mou_6                                          0.009095        NaN   
roam_ic_mou_7                                          0.003852        NaN   
roam_ic_mou_8                                          0.009626        NaN   
roam_og_mou_6                                          0.009181        NaN   
roam_og_mou_7                                          0.005140        NaN   
roam_og_mou_8                                          0.002256        NaN   
loc_og_t2t_mou_6                                       0.056315        NaN   
loc_og_t2t_mou_7                                       0.053847        NaN   
loc_og_t2t_mou_8                                       0.056163        NaN   
loc_og_t2m_mou_6                                       0.091990        NaN   
loc_og_t2m_mou_7                                       0.088011        NaN   
loc_og_t2m_mou_8                                       0.090425        NaN   
loc_og_t2f_mou_6                                       0.074701        NaN   
loc_og_t2f_mou_7                                       0.074100        NaN   
loc_og_t2f_mou_8                                       0.073670        NaN   
loc_og_t2c_mou_6                                      -0.005055        NaN   
loc_og_t2c_mou_7                                      -0.009558        NaN   
loc_og_t2c_mou_8                                       0.000979        NaN   
loc_og_mou_6                                           0.098194        NaN   
loc_og_mou_7                                           0.094071        NaN   
loc_og_mou_8                                           0.096021        NaN   
std_og_t2t_mou_6                                      -0.019218        NaN   
std_og_t2t_mou_7                                      -0.020933        NaN   
std_og_t2t_mou_8                                      -0.017158        NaN   
std_og_t2m_mou_6                                      -0.032855        NaN   
std_og_t2m_mou_7                                      -0.037804        NaN   
std_og_t2m_mou_8                                      -0.032901        NaN   
std_og_t2f_mou_6                                       0.031990        NaN   
std_og_t2f_mou_7                                       0.028310        NaN   
std_og_t2f_mou_8                                       0.033439        NaN   
std_og_t2c_mou_6                                            NaN        NaN   
std_og_t2c_mou_7                                            NaN        NaN   
std_og_t2c_mou_8                                            NaN        NaN   
std_og_mou_6                                          -0.034367        NaN   
std_og_mou_7                                          -0.038873        NaN   
std_og_mou_8                                          -0.032375        NaN   
isd_og_mou_6                                          -0.006340        N

In [29]:
# remove columns showing less variance

data_hvc=data_hvc.drop(['circle_id','loc_og_t2o_mou','std_og_t2o_mou','loc_ic_t2o_mou',
                        'last_date_of_month_6','last_date_of_month_7','last_date_of_month_8'],1)


In [30]:
#Converting datetime days

data_hvc['date_of_last_rech_6'] = data_hvc['date_of_last_rech_6'].astype('category')
data_hvc['date_of_last_rech_7'] = data_hvc['date_of_last_rech_7'].astype('category')
data_hvc['date_of_last_rech_8'] = data_hvc['date_of_last_rech_8'].astype('category')
data_hvc['date_of_last_rech_data_6'] = data_hvc['date_of_last_rech_data_6'].astype('category')
data_hvc['date_of_last_rech_data_7'] = data_hvc['date_of_last_rech_data_7'].astype('category')
data_hvc['date_of_last_rech_data_8'] = data_hvc['date_of_last_rech_data_8'].astype('category')
cat_columns = data_hvc.select_dtypes(['category']).columns
data_hvc[cat_columns] = data_hvc[cat_columns].apply(lambda x: x.cat.codes)

In [ ]:
# make a pairplot for the dervied variables created

sns.pairplot(data_hvc[[x for x in data_hvc.columns if 'difference' in x]])
plt.show()

In [ ]:
data_hvc.head()

## Imbalance check

In [ ]:
# Handle class imbalance

data_hvc['churn'].value_counts().plot(kind='bar').set_title('Churn')

## Train-Test split

In [ ]:
# change churn to numeric
data_hvc['churn'] = pd.to_numeric(data_hvc['churn'])

In [ ]:
y = data_hvc.churn
X = data_hvc.drop('churn', axis=1)

# setting up testing and training sets
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=4 , stratify=y)

In [ ]:
print(X_train.shape)

In [ ]:
X_train.columns

# PCA

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

pca = Pipeline([('scaler', StandardScaler()), ('pca', PCA())])

In [ ]:
pca.fit(X_train)
churn_pca = pca.fit_transform(X_train)

In [ ]:
# extract pca model from pipeline
pca = pca.named_steps['pca']

In [ ]:
#Making the screeplot - plotting the cumulative variance against the number of components
%matplotlib inline
fig = plt.figure(figsize = (12,8))
plt.plot(np.cumsum(pca.explained_variance_ratio_))
plt.xlabel('number of components')
plt.ylabel('cumulative explained variance')
plt.show()

at around 60 components about 90% of the variance is expplained

# PCA and Logistic regression

In [ ]:
# create pipeline
from sklearn.linear_model import LogisticRegression

PCA_VARS = 60
steps = [('scaler', StandardScaler()),
         ("pca", PCA(n_components=PCA_VARS)),
         ("logistic", LogisticRegression(class_weight='balanced'))
        ]
pipeline = Pipeline(steps)

In [ ]:
# fit model
pipeline.fit(X_train, y_train)

# check score on train data
pipeline.score(X_train, y_train)

In [ ]:
# predict churn on test data
y_pred = pipeline.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)


In [ ]:
TN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
TP=cm[1,1]

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

In [ ]:
Sensitivity=TP/(TP+FN)
Specificity=TN/(TN+FP)
print('Sensitivity=',Sensitivity)
print('Specificity=',Specificity)
# check area under curve
y_pred_prob = pipeline.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

## Hyperparameter Tuning

In [ ]:
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import GridSearchCV


In [ ]:
# PCA
pca = PCA()

# logistic regression - the class weight is used to handle class imbalance - it adjusts the cost function
logistic = LogisticRegression(class_weight={0:0.1, 1: 0.9})

# create pipeline
steps = [("scaler", StandardScaler()), 
         ("pca", pca),
         ("logistic", logistic)
        ]

# compile pipeline
pca_logistic = Pipeline(steps)

# hyperparameter space
params = {'pca__n_components': [60, 80],
          'logistic__C': [0.1, 0.5, 1, 2, 3, 4, 5, 10], 
          'logistic__penalty': ['l1', 'l2']}

# create 5 folds
folds = StratifiedKFold(n_splits = 5, 
                        shuffle = True, 
                        random_state = 4)

# create gridsearch object
model = GridSearchCV(estimator=pca_logistic,
                     cv=folds, 
                     param_grid=params, 
                     scoring='roc_auc', 
                     n_jobs=-1, 
                     verbose=1)

In [ ]:
# fit model

model.fit(X_train, y_train)

In [ ]:
# print best hyperparameters

print("Best AUC: ", model.best_score_)
print("Best hyperparameters: ", model.best_params_)

In [ ]:
# predict churn on test data

y_pred = model.predict(X_test)

# create onfusion matrix

cm = confusion_matrix(y_test, y_pred)
print(cm)

In [ ]:
TN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
TP=cm[1,1]

In [ ]:
Sensitivity=TP/(TP+FN)
Specificity=TN/(TN+FP)
print('Sensitivity=',Sensitivity)
print('Specificity=',Specificity)
# check area under curve
y_pred_prob = pipeline.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

# Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
# random forest - the class weight is used to handle class imbalance - it adjusts the cost function
forest = RandomForestClassifier(class_weight={0:0.1, 1: 0.9}, n_jobs = -1)

# hyperparameter space
params = {"criterion": ['gini', 'entropy'], 
          "max_features": ['auto', 0.4]}

# create 5 folds
folds = StratifiedKFold(n_splits = 5, 
                        shuffle = True,
                        random_state = 4)

# create gridsearch object
model_1 = GridSearchCV(estimator=forest, 
                     cv=folds, 
                     param_grid=params,
                     scoring='roc_auc',
                     n_jobs=-1, 
                     verbose=1)

In [ ]:
# fit model

model_1.fit(X_train, y_train)

In [ ]:
# print best hyperparameters
print("Best AUC: ", model_1.best_score_)
print("Best hyperparameters: ", model_1.best_params_)

In [ ]:
# predict churn on test data
y_pred = model_1.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

In [ ]:
TN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
TP=cm[1,1]

In [ ]:
Sensitivity=TP/(TP+FN)
Specificity=TN/(TN+FP)
print('Sensitivity=',Sensitivity)
print('Specificity=',Specificity)
# check area under curve
y_pred_prob = pipeline.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

This is a poor model

# Random Forest with best features

In [ ]:
# run a random forest model on train data
max_features = int(round(np.sqrt(X_train.shape[1])))    # number of variables to consider to split each node
print(max_features)

rf_model = RandomForestClassifier(n_estimators=100, 
                                  max_features=max_features,
                                  class_weight={0:0.1, 1: 0.9}, 
                                  oob_score=True, 
                                  random_state=4, 
                                  verbose=1)

In [ ]:
# fit model
rf_model.fit(X_train, y_train)

In [ ]:
# OOB score
rf_model.oob_score_

In [ ]:
# predict churn on test data
y_pred = rf_model.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

In [ ]:
TN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
TP=cm[1,1]

In [ ]:
Sensitivity=TP/(TP+FN)
Specificity=TN/(TN+FP)
print('Sensitivity=',Sensitivity)
print('Specificity=',Specificity)
# check area under curve
y_pred_prob = pipeline.predict_proba(X_test)[:, 1]
print("AUC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

still not good , we now check by taking only the best features

# feature_importance_

In [ ]:
# predictors
features = data_hvc.drop('churn', axis=1).columns

# feature_importance
importance = rf_model.feature_importances_

# create dataframe
feature_importance = pd.DataFrame({'variables': features, 'importance_percentage': importance*100})
feature_importance = feature_importance[['variables', 'importance_percentage']]

# sort features
feature_importance = feature_importance.sort_values('importance_percentage', ascending=False).reset_index(drop=True)
print("Sum of importance=", feature_importance.importance_percentage.sum())
feature_importance

In [ ]:
# extract top 'n' features
top_n = 20
top_features = feature_importance.variables[0:top_n]

In [ ]:
# plot feature correlation
import seaborn as sns
plt.rcParams["figure.figsize"] =(20,10)
sns.heatmap(data=X_train[top_features].corr(),annot=True)

In [ ]:
top_features = ['loc_ic_t2t_mou_8','total_ic_mou_8','loc_og_t2t_mou_8','roam_og_mou_8','loc_ic_t2m_mou_8',
               'max_rech_amt_8','roam_ic_mou_8','av_rech_amt_data_8','std_ic_mou_8']
X_train = X_train[top_features]
X_test = X_test[top_features]

In [ ]:
# logistic regression
steps = [('scaler', StandardScaler()), 
         ("logistic", LogisticRegression(class_weight={0:0.1, 1:0.9}))
        ]

# compile pipeline
logistic = Pipeline(steps)

# hyperparameter space
params = {'logistic__C': [0.1, 0.5, 1, 2, 3, 4, 5, 10], 'logistic__penalty': ['l1', 'l2']}

# create 5 folds
folds = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 4)

# create gridsearch object
model = GridSearchCV(estimator=logistic, cv=folds, param_grid=params, scoring='roc_auc', n_jobs=-1, verbose=1)

In [ ]:
# fit model
model.fit(X_train, y_train)

In [ ]:
# print best hyperparameters
print("Best AUC: ", model.best_score_)
print("Best hyperparameters: ", model.best_params_)

In [ ]:
# predict churn on test data
y_pred = model.predict(X_test)

# create onfusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

In [ ]:
TN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
TP=cm[1,1]

In [ ]:
Sensitivity=TP/(TP+FN)
Specificity=TN/(TN+FP)
print('Sensitivity=',Sensitivity)
print('Specificity=',Specificity)
y_pred_prob = model.predict_proba(X_test)[:, 1]
print("ROC:    \t", round(roc_auc_score(y_test, y_pred_prob),2))

In [ ]:
logistic_model = model.best_estimator_.named_steps['logistic']

In [ ]:
# intercept
intercept_df = pd.DataFrame(logistic_model.intercept_.reshape((1,1)), columns = ['intercept'])

In [ ]:
# coefficients
coefficients = logistic_model.coef_.reshape((9, 1)).tolist()
coefficients = [val for sublist in coefficients for val in sublist]
coefficients = [round(coefficient, 3) for coefficient in coefficients]

logistic_features = list(X_train.columns)
coefficients_df = pd.DataFrame(logistic_model.coef_, columns=logistic_features)

In [ ]:
# concatenate dataframes
coefficients = pd.concat([intercept_df, coefficients_df], axis=1)
coefficients

### Reduction in STD and ISD prices will help the company


### The company also need to focus on the incoming and outgoing call prices within the operator and outside the ### operator as well


